### ISSUE TITLE
 Feature request - sorted data for map_script

### ISSUE THREAD

**Issue by jeacott1** [elastic/elasticsearch#25923]
it would be extremely handy if scripted_metric map_script could be configured to sort the incoming data by something other than the native id.

if the incoming data could be sorted by an arbitrary date field for example, then reimann sums and similar calculations can be made very efficiently. Would be the simplest way to get an efficient area under graph calc with minimum fuss.

**Comment by danielmitterdorfer** [comment]
@colings86 / @polyfractal do you have any thoughts on this?

**Comment by nik9000** [comment]
This is similar to the sorted aggs work that @jimczi has been talking about for a few months. I believe it isn't super simple to get the hits in any order other than the one that they come in now.

**Comment by polyfractal** [comment]
Yeah, that was my thought as well.  I wonder if existing index sorting (https://www.elastic.co/guide/en/elasticsearch/reference/master/index-modules-index-sorting.html) would help.  It'd guarantee the order within a segment, but I'm not sure how it would interact with aggs when traversing segments.  My guess is that the script would have to handle going "backwards" when a new segment is loaded/collected. But I haven't actually played with index sorting yet so dunno.

I believe @jimczi's idea was to build on top of index sorting, but I'm not sure what else is needed to accomplish it. :)

**Comment by jeacott1** [comment]
what if the map interface was adjusted so that at least per segment could be accessed in order? it wouldn't be perfect but it would be a big step forward and likely more than close enough for my needs.

**Comment by jeacott1** [comment]
@polyfractal well thats put a dent in my day :(
"discretize the data into some acceptable level of granularity" does not at all deal with the issue of calculating an integral, or many other custom functions over a timeseries. 
very sad to see this shelved as its a huge blocker for a lot of timeseries usecases, indeed anything vaguely custom needs access to sorted data.

**Comment by polyfractal** [comment]
Sorry to be the bearer of bad news :(

Some thoughts:

> "discretize the data into some acceptable level of granularity" does not at all deal with the issue of calculating an integral, or many other custom functions over a timeseries.

Assuming we're working with data points stored in Elasticsearch (e.g. not a real-valued function itself, but the output from that function at discrete points), the data has already been discretized based on the sampling period or timestamp granularity.  E.g. if your timestamps are millisecond precision, you're implicitly discretized to millisecond approximations of the real signal.  Calculating the integral on any dataset of discrete points is automatically an approximation of the true integral.

With that in mind, there are some options:

- You could do with a sorted scroll and iterate over the raw data points, calculating the integral locally.  It could be made faster by using index-time sorting.
- Another option is to use the new composite aggregation and set a `date_histo` to the same interval as your sampling period, which will give you a pageable view into the time-series data (without the memory hit of a regular aggregation).  Index sorting could help here too.
- Finally, if you have enough memory, you could just do a regular `date_histo` with the interval set to the sampling period.

The "gotcha" with the those options is that they are all working effectively on the raw data.  There's no real way to work with sorted, sequential data efficiently in a distributed map/reduce environment.  Since each shard is independent of the other shards, any calculation that requires sequential access to a sorted stream can't be done in a shard-local manner.  No individual shard knows if it has a complete view of the timeseries at a point in time.

So if parallelism is needed, best a map/reduce system can do is summarize the data into buckets like `date_histo` does, because that can be done in a shard-local manner and reduced to the final set of buckets.  Then you work on that summarized, larger granularity view instead.  For the intergal problem, the ∑<sub>[k=0, n]</sub>(bucket_count<sub>k</sub> * bucket_width<sub>k</sub>) is an approximation of the integral.

There are some slightly more accurate ways to calculate an integral in a map/reduce environment, such as doing [Monte Carlo approximations](https://hplgit.github.io/map-reduce4sci/doc/pub/MapReduce.html).  But my understanding is that this requires chunks of sequential data to be co-located (which Elasticsearch doesn't provide), and it's still an approximation.

I'd really love to support more advanced time-series functionality, but they become very tricky in a distributed environment.  Many time-series calculations require sequential data which is not something that's doable in a map/reduce situation. Simply sorting each shard independently doesn't help because we still have to deal with the results from other shards.

_Addendum_ 
There is a scenario where the user has configured custom routing, so that each independent time-series is routed to a specific shard.  In that case we could theoretically provide a sorted stream of data, since we know the data for the entire series is shard-local.  I doubt we'd ever build features like that though, since it essentially gives up on scalability, would create a certain class of aggs that only work if your data is configured a certain way, and would not work with multiple non-colocated series (e.g. correlate two time-series that were on different shards)

**Comment by jeacott1** [comment]
thanks for the huge answer!
re shards and reading data, if I use a single shard per index (and/or use routing affinity), and manage data distribution by creating new sequential (in time) indexes, and replicas for read scale, I think I can have my cake and eat it, each shard would contain a complete set of data for a time period, and a map reduce across indexes would be straight forward. 
I understand that a sorted map_script input would not be useful in every circumstance, but I think just because its not directly useful in the common 'doc hashed to multiple shards case' is no reason to dismiss the option. if that were the case, shard routing would likely never have made the cut either.

hoping this gets further consideration at some point. deferring to scrolling api and hauling all the data out of ES doesn't really sound like a viable option. 

Cheers.
